File: javascript/perfect-numbers/perfect-numbers.js
//
// This is only a SKELETON file for the 'Perfect Numbers' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const classify = (number) => {
  if (number <= 0) throw new Error('Classification is only possible for natural numbers.');

  let sum = 0;
  for (let f = 1; f < number; f++) {
    if (number % f === 0) {
      sum += f;
    }
  }

  if (number === sum) return 'perfect';
  if (number < sum) return 'abundant';
  return 'deficient';
};

File: zig/perfect-numbers/perfect_numbers.zig
const std = @import("std");

pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

/// Asserts that `n` is nonzero.
pub fn classify(n: u64) Classification {
    std.debug.assert(n > 0);

    var sum: u64 = 0;
    for (1..n) |f| {
        if (n % f == 0) {
            sum += f;
        }
    }

    return if (n == sum) Classification.perfect else if (n < sum) Classification.abundant else Classification.deficient;
}
